modernizer-annotation-processor/src/main/java/org/gaul/modernizer_annotation_processor/ModernizerAnnotationProcessor.java

```diff
@@ -0,0 +1,331 @@
+        }
+    }
+
+    private File getOutputDirectory() {
+        try {
+            FileObject fileObjectToGetPath =
+                processingEnv.getFiler().createResource(
+                    StandardLocation.CLASS_OUTPUT,
+                    "",
+                    "dummy-file.txt");
+            File outputDir =
+                new File(fileObjectToGetPath.getName()).getParentFile();
+            return ModernizerAnnotationOutput.getOutputDir(outputDir);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /* Return class header in a regex format
+     * to perform necessary checks in the plugin
+     */
+    private String getClassHeaderRegex(Element classElement) {
+        return getClassHeader(classElement).replace("$", "\\$") + "(\\$.+)?";
+    }
+
+    private String getClassHeader(Element classElement) {
+        String packageName = processingEnv
+            .getElementUtils()
+            .getPackageOf(classElement)
+            .getQualifiedName()
+            .toString()
+            .replace('.', '/');
+        String packagePrefix = !packageName.isEmpty() ? packageName + "/" : "";
+        String classHeader = packagePrefix + getFullClassName(classElement);
+        return classHeader;
+    }
+
+    private String getFullClassName(Element classElement) {
+        List<String> parentClasses = new ArrayList<String>();
+        Element enclosingElement = classElement.getEnclosingElement();
+        while (enclosingElement != null &&
+            enclosingElement.getKind().isClass()) {
+            parentClasses.add(enclosingElement.getSimpleName().toString());
+            enclosingElement = enclosingElement.getEnclosingElement();
+        }
+        StringBuilder className = new StringBuilder();
+        for (int index = parentClasses.size() - 1; index >= 0; index--) {
+            className.append(parentClasses.get(index));
+            className.append("$");
+        }
+        return className.toString() + classElement.getSimpleName();
+    }
+
+    private String getMethodIdentifierString(Element methodElement) {
+        TypeVisitor<ExecutableType, Void> executableTypeVisitor =
+            new SimpleTypeVisitor6<ExecutableType, Void>() {
+                @Override
+                public ExecutableType visitExecutable(
+                    ExecutableType executableType, Void obj
+                ) {
+                    return executableType;
+                }
+            };
+        ExecutableType method =
+            methodElement.asType().accept(executableTypeVisitor, null);
+        List<? extends TypeMirror> methodParams = method.getParameterTypes();
+
+        /* Fetching the return type of the method in the format ASM
+         * parses the return type in a method descriptor.
+         */
+        TypeMirror returnType = method.getReturnType();
+        String returnTypeString = "";
+        if (returnType.getKind().equals(TypeKind.ARRAY)) {
```

Comment:
This logic is repeated. It would make more sense to have one method like:
```java
private String getRep(TypeMirror type) {
  if (returnType.getKind().equals(TypeKind.ARRAY)) {
    // array stuff
    return getRep(paramType) + Strings.repeat("[]", arrayLength);
  } else if (param.getKind().equals(TypeKind.TYPEVAR)) {
    return Object.class.getName();
  } else {
    String rep = type.toString();
    int ltIndex = rep.indexOf("<");
    return ltIndex == -1 ? rep : rep.substring(0, ltIndex);
  }
}
```
This would also help with the fact that the `getParameter` method is really just applied to both parameter and return types.